analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs

```diff
@@ -148,15 +161,22 @@ declaration switch
             recursive switch
             {
                 _ when valueConstraint == ObjectConstraint.Null => BoolConstraint.False,
+                _ when RecursivePatternAlwaysMatchesAnyNotNull(recursive) => BoolConstraint.True,
+                _ => null,
+            };
+
+        private static bool RecursivePatternAlwaysMatchesAnyNotNull(IRecursivePatternOperationWrapper recursive) =>
+            recursive switch
+            {
                 {
                     PropertySubpatterns.Length: 0,
                     DeconstructionSubpatterns.Length: 0,
-                } => BoolConstraint.True,
-                _ when SubPatternsAlwaysMatch(recursive) => BoolConstraint.True,
-                _ => null,
+                } => true,
```

Comment:
The type compatibility check is needed here too. Add some test cases for this. This is just a shortcut to prevent the expensive SubPatternsAlwaysMatches call. I think it is better to move this check into SubPatternsAlwaysMatches itself.

Reply:
The `SubPatternsAlwaysMatches` has a good name given what it does. I've prepended the type check, as it is needed always. That allows so simplify the syntax too